Spigot-Server-Patches/0422-Delay-chunk-unloads.patch

```diff
@@ -0,0 +1,194 @@
++        if (delayChunkUnloadsBy >= 0) {
++            net.minecraft.server.TicketType.DELAYED_UNLOAD.loadPeriod = delayChunkUnloadsBy;
++        }
++    }
+ }
+diff --git a/src/main/java/net/minecraft/server/ChunkMap.java b/src/main/java/net/minecraft/server/ChunkMap.java
+index 55f9f4e6e..c5b5f0d39 100644
+--- a/src/main/java/net/minecraft/server/ChunkMap.java
++++ b/src/main/java/net/minecraft/server/ChunkMap.java
+@@ -68,6 +68,7 @@ public abstract class ChunkMap extends LightEngineGraph {
+ 
+     protected abstract int b(long i);
+ 
++    public final void update(long coordinate, int level, boolean increaseInLevel) { this.b(coordinate, level, increaseInLevel); } // Paper - OBFHELPER
+     public void b(long i, int j, boolean flag) {
+         this.a(ChunkCoordIntPair.a, i, j, flag);
+     }
+diff --git a/src/main/java/net/minecraft/server/ChunkMapDistance.java b/src/main/java/net/minecraft/server/ChunkMapDistance.java
+index 8c1945687..4dc96eb66 100644
+--- a/src/main/java/net/minecraft/server/ChunkMapDistance.java
++++ b/src/main/java/net/minecraft/server/ChunkMapDistance.java
+@@ -30,7 +30,7 @@ public abstract class ChunkMapDistance {
+     private static final int b = 33 + ChunkStatus.a(ChunkStatus.FULL) - 2;
+     private final Long2ObjectMap<ObjectSet<EntityPlayer>> c = new Long2ObjectOpenHashMap();
+     public final Long2ObjectOpenHashMap<ArraySetSorted<Ticket<?>>> tickets = new Long2ObjectOpenHashMap();
+-    private final ChunkMapDistance.a e = new ChunkMapDistance.a();
++    private final ChunkMapDistance.a e = new ChunkMapDistance.a(); final ChunkMapDistance.a getTicketTracker() { return this.e; } // Paper - OBFHELPER
+     private final ChunkMapDistance.b f = new ChunkMapDistance.b(8);
+     private final ChunkMapDistance.c g = new ChunkMapDistance.c(33);
+     private final java.util.Queue<PlayerChunk> pendingChunkUpdates = new java.util.LinkedList<>(); // PAIL pendingChunkUpdates // Paper - use a queue
+@@ -52,16 +52,75 @@ public abstract class ChunkMapDistance {
+         this.m = executor1;
+     }
+ 
++    // Paper start - delay chunk unloads
++    private long nextUnloadId; // delay chunk unloads
++    private final Long2ObjectOpenHashMap<Ticket<Long>> delayedChunks = new Long2ObjectOpenHashMap<>();
++    public final void removeTickets(long chunk, TicketType<?> type) {
++        ArraySetSorted<Ticket<?>> tickets = this.tickets.get(chunk);
++        if (tickets == null) {
++            return;
++        }
++        if (type == TicketType.DELAYED_UNLOAD) {
++            this.delayedChunks.remove(chunk);
++        }
++        boolean changed = tickets.removeIf((Ticket<?> ticket) -> {
++            return ticket.getTicketType() == type;
++        });
++        if (changed) {
++            this.getTicketTracker().update(chunk, getLowestTicketLevel(tickets), false);
++        }
++    }
++
++    private final java.util.function.LongFunction<Ticket<Long>> computeFuntion = (long key) -> {
```

Comment:
This probably needs a better name?

Reply:
It's not even spelled correctly, looks like I meant to name it but forgot